### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by hsxyl**

### CODE REVIEW

File: nep141-token-convertor-contract/src/account.rs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        let balance = *self
+            .tokens
+            .get(token_id)
+            .expect("Fail to withdraw nonexistent token.");
+        self.tokens.remove(token_id);
+        return balance;
+    }
+
+    pub fn storage_usage(&self) -> u64 {
+        INIT_ACCOUNT_STORAGE + self.tokens.len() as u64 * (ACC_ID_AS_KEY_STORAGE + U128_STORAGE)
```
Comment (riversyang) [inline nep141-token-convertor-contract/src/account.rs:62]: I think the storage usage should be `U128_STORAGE +  + self.tokens.len() as u64 * (ACC_ID_AS_KEY_STORAGE + U128_STORAGE)`.

File: nep141-token-convertor-contract/src/token_receiver.rs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                    token_id, convert_action.input_token_id
+                );
+                assert_eq!(
+                    amount, convert_action.input_token_amount,
+                    "transferred token amount: {} not eq convert_action token amount: {} ",
+                    amount.0, convert_action.input_token_amount.0
+                );
+                let (receive_token_id, receive_token_amount) =
+                    self.internal_convert(convert_action.pool_id, &token_id, amount.0);
+                self.internal_send_tokens(&sender_id, &receive_token_id, receive_token_amount);
```
Comment (riversyang) [inline nep141-token-convertor-contract/src/token_receiver.rs:65]: The function `internal_send_tokens` needs 30T gas to complete. But the prepaid gas for function `ft_transfer_call` is set to 25T in fungible token contract template. We need to double check whether the function `ft_transfer_resolved` can be executed successfully.

File: nep141-token-convertor-contract/src/conversion_pool.rs [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+            } else {
+                pool.withdraw_out_token(amount.0);
+            }
+        });
+        // pool should finish withdraw here
+        self.internal_send_tokens(&env::predecessor_account_id(), &token_id, amount.0);
+    }
+
+    #[payable]
+    fn delete_pool(&mut self, pool_id: PoolId) {
```
Comment (riversyang) [inline nep141-token-convertor-contract/src/conversion_pool.rs:357]: There are 3 possible promises which can be created in this function. I'm not sure what will happen if the prepaid gas can only fulfill the first one or two promise(s). Need to double check.
I recommend to check the prepaid gas before actual process.

**Review by riversyang — CHANGES_REQUESTED** [review]
I pointed out a few tricky parts in contract. And the tests failed in my computer because of compile errors. Need to double check.

File: nep141-token-convertor-contract/src/account.rs (hunk 1 above)
Comment (hsxyl) [reply nep141-token-convertor-contract/src/account.rs:62]: I think both `U128_STORAGE` and  `INIT_ACCOUNT_STORAGE `  are ok.Just make sure size of increasing storage when adding an account  equals size of decreasing when remove an account.

File: nep141-token-convertor-contract/src/token_receiver.rs (hunk 2 above)
Comment (hsxyl) [reply nep141-token-convertor-contract/src/token_receiver.rs:65]: I have add a check of gas in `internal_send_token` which requiring remind gas must more than `ft_transfer` and `ft_transfer_resolved` cost.

File: nep141-token-convertor-contract/src/conversion_pool.rs (hunk 3 above)
Comment (hsxyl) [reply nep141-token-convertor-contract/src/conversion_pool.rs:357]: I have changed logic in `delete_pool`.Now requiring withdraw all `in_token` and `out_token` in pool before deleting pool.And admin can also execute withdraw token for creator now.

**Review by riversyang — APPROVED** [review]

**Pull request merged**